dspy/teleprompt/bootstrap_finetune.py

```diff
@@ -213,22 +225,74 @@ def bootstrap_trace_data(
         num_threads=num_threads,
         display_progress=True,
         return_outputs=True,
-        provide_traceback=True,  # TODO(check with team)
+        provide_traceback=False,  # TODO(check with team)
+        max_errors=len(dataset)*10,  # TODO(check with team)
+        failure_score=failure_score,
     )
 
     def wrapped_metric(example, prediction, trace=None):
         prediction, _ = prediction
+        if isinstance(prediction, FailedPrediction):
+            return prediction.format_reward or format_failure_score
         return metric(example, prediction, trace) if metric else True
 
     def wrapped_program(**kwargs):
         with dspy.context(trace=[]):
-            return program(**kwargs), dspy.settings.trace.copy()
+            try:
+                return program(**kwargs), dspy.settings.trace.copy()
+            except ValueError as ve:
+                if len(ve.args) in [4, 6] and "Failed to parse response as per signature from original completion with input" in ve.args[0] and capture_failed_parses:
```

Comment:
This except block looks pretty hacky, is the motivation getting the failed predictions on parse errors? If so, we can make a custom exception `dspy.exception.ParseError` and capture it here.

Reply:
Yes, that sounds much better! I wanted to make minimal changes to the rest of the dspy, hence opted to include all as args, but totally agree this is hacky, and defining a custom exception is better.